src/ipred_prepare.rs

```diff
@@ -120,59 +118,72 @@ static av1_intra_prediction_edges: [av1_intra_prediction_edge; N_IMPL_INTRA_PRED
     b
 };
 
-pub unsafe fn rav1d_prepare_intra_edges<BD: BitDepth>(
+pub fn rav1d_prepare_intra_edges<BD: BitDepth>(
     x: c_int,
-    have_left: c_int,
+    have_left: bool,
     y: c_int,
-    have_top: c_int,
+    have_top: bool,
     w: c_int,
     h: c_int,
     edge_flags: EdgeFlags,
     dst: *const BD::Pixel,
     stride: ptrdiff_t,
-    prefilter_toplevel_sb_edge: *const BD::Pixel,
+    prefilter_toplevel_sb_edge: Option<&[BD::Pixel]>,
     mut mode: IntraPredMode,
-    angle: *mut c_int,
+    angle: &mut c_int,
     tw: c_int,
     th: c_int,
     filter_edge: c_int,
-    topleft_out: *mut BD::Pixel,
+    topleft_out: &mut [BD::Pixel],
+    topleft_origin: usize, // position of top-left sample in `topleft_out`
```

Comment:
```suggestion
    topleft_out_offset: usize,
```

I think `offset` makes more sense here than `origin`, since `offset` is often used in this way.  We do have a `Cursor` type @randomPoison added, but this is fine, too, and the logic is more explicit so it's easy to follow the changes.